Add package archive fallback mechanism

d12frosted opened the pull request:
Seeing all these issues with package installation due to MELPA availability problems I decided to improve status check for all elpa archives and error about archive being unavailable instead of showing errors like 'package xxx is not available' in the future.

Also I removed timeout heuristic as it's not 100% accurate. I don't see any reason for it to be here once `configuration-layer-check-archive-status` is used. 

Note that `configuration-layer-check-archive-status` is a little bit beefed (more than we need at this point), but it might become handy in future. 

Oh, and I am not sure about the error message though. @syl20bnr, you might want to improve it. 😸 

P. S. Hope I didn't miss anything. 

P. P. S. Somewhat fixes #6836. Another related issue is #4453.

TheBB commented on core/core-configuration-layer.el:
```diff
@@ -1953,6 +1939,26 @@ FILE-TO-LOAD is an explicit file to load after the installation."
+               url)
+      (condition-case nil
+          (if (url-http-file-exists-p url)
+              (setq state 'avaialble)
```
available

d12frosted replied:
My typing skills are weak 👀 Thanks for noticing this typo!

TheBB commented on core/core-configuration-layer.el:
```diff
@@ -1953,6 +1939,26 @@ FILE-TO-LOAD is an explicit file to load after the installation."
+            (setq state 'unavailable))
+        ((error) (setq state 'unavailable)))
+      (setq configuration-layer--elpa-archives-status (delq obj configuration-layer--elpa-archives-status))
+      (add-to-list 'configuration-layer--elpa-archives-status (cons archive state)))
```
You can do

    (setcdr (assq archive configuration-layer--elpa-archives-status) state)

d12frosted replied:
Wow. Didn't know about `setcdr`. Thanks. Though in this case you have to use `assoc` instead of `assq` as we are dealing with strings.

TheBB commented:
If `dotspacemacs-elpa-timeout` isn't needed any more you should remove it (and the tests for it).

d12frosted commented:
I have removed `dotspacemacs-elpa-timeout` as well as related test. Also I have updated error message in order to make it more clear.

d12frosted commented:
Just pushed another commit that adds fallback functionality when one of the archives is down. You can test it by returning `'unavailable` in `configuration-layer-check-archive-status` for any archive of your liking. For example,

```
(when (string-equal archive "melpa")
      (setq state 'unavailable))
```

So when any archive is not available, `configuration-layer/retrieve-package-archives` will fallback to `configuration-layer-fallback-package-archive` for unavailable archive. 

As fallback I am using snapshot provided by @zilongshanren. Thanks to https://github.com/melpa/melpa/issues/4156 I've found this nice 'mirror'. The good thing is that MELPA is going to create official mirror that people could use when MELPA is down. 

P. S. I've made two commits for now - gonna squash them later on demand. Just wanna keep first commit for now.

d12frosted commented on core/core-configuration-layer.el:
```diff
@@ -201,6 +198,19 @@ LAYER has to be installed for this method to work properly."
+    ("gnu" . unknown))
+  "Status of ELPA archives.")
+
+(defvar configuration-layer-fallback-package-archive
```
I am planning to convert it list of fallback archives.

d12frosted commented on core/core-configuration-layer.el:
```diff
@@ -329,24 +337,16 @@ refreshed during the current session."
-                         'error))))
+        (if (eq 'unavailable
+                (configuration-layer-check-archive-status (car archive)))
+            (setcdr
```
Here I am going to check fallback archives as well and if all of them unavailable - error like in first commit.

d12frosted commented:
Pushed another commit. 

Instead of using one fallback url, use package-archives compatible list of fallback urls. If currently set url is not reachable, then try to connect to one of fallback urls. If all urls are unavailable - show an error with some hints and explanations.

Again, you might want to change error message. But I tried to make it as useful as possible. 😸

d12frosted commented:
And a good thing, now https://github.com/syl20bnr/spacemacs-elpa is used as primary fallback archive 😸

d12frosted commented on core/core-configuration-layer.el:
```diff
@@ -201,6 +198,26 @@ LAYER has to be installed for this method to work properly."
+    ("gnu" . unknown))
+  "Status of ELPA archives.")
+
+(defvar configuration-layer-fallback-package-archives
```
I wonder if we really need this. I mean, fallback archives can be added to the end of `configuration-layer--elpa-archives`, so instead of switching to values from `configuration-layer-fallback-package-archives` we can just remove first associated url and search for another. Once no urls are available for specific archive - error!

d12frosted commented on core/core-configuration-layer.el:
```diff
@@ -295,41 +299,61 @@ cache folder.")
+        (state 'unavailable))
+    (condition-case nil
+        (when (or (and (string-prefix-p "/" url)
+                       (file-readable-p url))
```
Not sure that this is the best way of checking that archive url is accessible. `file-accessible-directory-p`? I haven't tested situations when `archive-contents` file is readable, but all other files are not readable by current user.

syl20bnr replied:
@d12frosted Where can we find the documentation about `file-readable-p` capable of checking URLs except in the C source ? This function does not block for a long time when the URL is not reachable ? What kind of tests did you do to check that `file-readable-p` is not blocking for a long time ?

syl20bnr replied:
> What kind of tests did you do to check that file-readable-p is not blocking for a long time ?

I mean, did you test regular timeout method and `file-readable-p` with the same input ?

d12frosted commented:
Ok, I've made several changes. I simplified fallback mechanism, removed two redundant variables and improved error message. Oh and also I made it possible to use local package archive urls.

d12frosted commented:
I spammed pretty much in this PR with commits. I'll squash them once @syl20bnr give the green light 😸 For now I wish to preserve history. 😸

ninrod commented:
This should close #7103 and provide a way for users behind corporate firewalls to use spacemacs in all it's glory.

d12frosted commented:
@ninrod is GitHub available for you behind your firewall?

ninrod commented:
@d12frosted yes it is. We have a severely slowed down connection pipe to and from it though. So the local clone is like heaven.

But really, the local melpa solution is so awesome that I'm using it in home, even without a firewall. 
With a single git clone you have the whole melpa, elpa and what not in a local folder that you control. No connection hiccups, I don't have to worry about melpa going down, reconfigurations are like instantaneous, I can tag a specific snapshot that I know for a fact that works with my config. What more can I ask? That should be the default. So robust, efficient, fast, configurable. Awesome.

d12frosted commented:
@ninrod I asked about GitHub just because some of packages used by Spacemacs are downloaded using `quelpa` from GitHub directly. This fallback mechanism doesn't cover such packages. E. g. when GitHub is down (?!) it's impossible to start Spacemacs from scratch even if you have local copies of elpa package archives (namely GNU ELPA, MELPA, Stable MELPA and Org-mode package archive). I know that I already said this in other thread, but I just want to make it clear for others as well. 

> reconfigurations are like instantaneous

Indeed. Also it makes bootstrap testing much easier. 

> I can tag a specific snapshot that I know for a fact that works with my config

And this as well. Might be very useful for corporations, where they want to use specific stable version of Spacemacs with specific package archives snapshot (so no upstream changes break flow). 

> That should be the default. So robust, efficient, fast, configurable. Awesome.

Glad you like it 💃

ninrod commented:
@d12frosted I should mention that I have the option of mirroring specific github repos inside an internal git server behind corp firewall. Out of curiosity, am I able to change those quelpa URLs to match my specific internal mirrors?

ninrod commented:
let's do this!
